### PR TITLE
Feature: `variant2`

### PR DESCRIPTION
**Pull request by Clueliss**
Adds `variant2` a more efficient `std::variant` implementation for exactly two elements.
Should be a drop in replacement for `std::variant<T, U>` if I did not overlook anything.

POBR version increased because the layout of `polymorphic_allocator` with 2 allocators changed. But.. we are not using polymorphic_allocator anymore, so not POBR break?

TODO:
- [x] add (unit) tests, is already indirectly tests through polymorphic allocator
- [x] add example
- [x] update readme
- [x] determine if we consider this a POBR break

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -6,7 +6,7 @@ project(
   DESCRIPTION
     "This template library is a collection of template-oriented code that we, the Data Science Group at UPB, found pretty handy. It contains: `switch_cases` (Use runtime values in compile-time context), `integral_template_tuple` (Create a tuple-like structure that instantiates a template for a range of values), `integral_template_variant` (A wrapper type for `std::variant` guarantees to only contain variants of the form `T<IX>` and `for_{types,values,range}` (Compile time for loops for types, values or ranges))."
   HOMEPAGE_URL "https://dice-research.org/")
-set(POBR_VERSION 1)  # Persisted Object Binary Representation
+set(POBR_VERSION 2)  # Persisted Object Binary Representation
```
Comment (Clueliss) [inline CMakeLists.txt:9]: TODO figure out if we care that the layout of `polymorphic_allocator` changed. I would assume we don't

Comment (bigerl) [reply CMakeLists.txt:9]: Let's keep the version and remove polymorphic_allocator from the tracked files for POBR.

**Review by mcb5637 — CHANGES_REQUESTED** [review]
overall, looks good, just a few small issues.

File: include/dice/template-library/variant2.hpp [hunk 2]
```diff
@@ -0,0 +1,876 @@
...
+                            a_ = other.a_;
+                            break;
+                        }
+                        case discriminant_type::Second: {
+                            try {
+                                a_.~T();
+                                new (&b_) U{other.b_};
+                            } catch (...) {
+                                discriminant_ = discriminant_type::ValuelessByException;
+                                std::rethrow_exception(std::current_exception());
```
[the remark this answers is not in the document]
Comment (Clueliss) [reply include/dice/template-library/variant2.hpp:314]: I was not aware of this, thx

**Review by mcb5637 — APPROVED** [review]

**Pull request merged**